feat: install and upgrade instructions for management cluster in Kubeflow 1.2

Bobgy opened the pull request:
/assign @chensun

chensun commented on management/README.md:
````diff
@@ -4,14 +4,90 @@ This directory contains the configuration needed to setup a management GKE clust
+1. Check your existing config connector version:
+    ```bash
+    # For Kubeflow 1.1, it should be 1.15.0
+    $ kubectl get namespace cnrm-system -ojsonpath='{.metadata.annotations.cnrm\.cloud\.google\.com\/version}'
````
Was `cnrm-system` set somewhere?
Trying this command in my setup, I don't have this namespace.

Bobgy replied:
It's in the management cluster

chensun replied:
I see. Thanks!

chensun commented on management/README.md:
````diff
@@ -4,14 +4,90 @@ This directory contains the configuration needed to setup a management GKE clust
+    kubectl delete validatingwebhookconfiguration validating-webhook.cnrm.cloud.google.com --ignore-not-found --wait=true
+    kubectl delete mutatingwebhookconfiguration mutating-webhook.cnrm.cloud.google.com --ignore-not-found --wait=true
+    ```
+    These commands uninstall the config connector without removing your resources.
````
>     These commands uninstall the config connector without removing your resources.

The resources here refer to the cluster, right?
What about the `<NAME>-cnrm-system@<PROJECT_ID>.iam.gserviceaccount.com` SA? Should we delete it as part of the uninstallation process?

Bobgy replied:
Good point! Let's add it

Bobgy replied:
Added

chensun commented on tests/management-kf-1.1/README.md:
````diff
@@ -0,0 +1,14 @@
+Usage:
+```bash
+# deploy management cluster in Kubeflow 1.1.0
+make test
````
nit: While I understand this folder is for testing, `make test` seems a bit odd for the purpose of `deploy management cluster in Kubeflow 1.1.0`

Bobgy replied:
It's also creating a gcs bucket resource though. Do you have any suggestions?

chensun replied:
I don't really have a good suggestion, just feel like the words `test` and `deployment` are not close.
Maybe something like `make test-setup` or `make create-test-cluster`?

Bobgy replied:
PTAL, I've refactored tests to leverage some ways to dynamically set variables for makefile

chensun commented on tests/management-kf-1.1/storagebucket.yaml:
```diff
@@ -0,0 +1,12 @@
+metadata:
+  annotations:
+    cnrm.cloud.google.com/project-id : gongyuan-pipeline-test
+  name: gongyuan-test-5
```
> gongyuan-pipeline-test

nit: TODO: use a generic name?

Bobgy replied:
SGTM

chensun commented on tests/management/Makefile:
```diff
@@ -10,8 +10,11 @@ PROJECT=gongyuan-pipeline-test
+
 # Test 1.1.0
-MANAGEMENT_URL=https://github.com/kubeflow/gcp-blueprints.git/management@v1.1.0
+# MANAGEMENT_URL=https://github.com/kubeflow/gcp-blueprints.git/management@1.1.0
```
Give this a separate name, like `MANAGEMENT_1_1_0_URL` ? May need to parameterize the URL.

Bobgy replied:
If we uncomment this, then we can test 1.1.0, so it has to be the same name

chensun replied:
Understand, just a low priority thinking that if we can parameterize this as if it were an input to a function, then we can do `make test-master` or `make test-kf-1-1`, and it will pick up the right URI without the need that someone manually comments/uncomments the code.

Bobgy commented on management/README.md:
````diff
@@ -4,14 +4,90 @@ This directory contains the configuration needed to setup a management GKE clust
+    ```
+1. Set values for the name, project, and location of your management cluster:
+    ```bash
+    kpt cfg set ./instance name NAME
````
Great point!
Let's do what you said

Bobgy commented on management/Makefile:
```diff
@@ -4,13 +4,22 @@
 # LOCATION: Location of the management cluster. You can choose either regional or zonal.
 # PROJECT: Google Cloud project where this management cluster is created in.
+#
+# You can set these values by
```
I agree with you after trying the setup, it's a lot harder to get my head around these commands than a settings.yaml.

Bobgy replied:
/cc @jlewi @subodh101

What do you think?

Bobgy replied:
FYI, previous change: https://github.com/kubeflow/gcp-blueprints/pull/137

jlewi replied:
I think using kpt consistently is preferable as this creates a consistent experience. 
Users can do `kpt cfg list-setters .` to see all of the setters and then set them appropriately. 
This is consistent with respect to how arbitrary kpt packages are managed.
e.g. if a user downloads some new kpt package the expectation would be they run list-setters to see what setters exist and then set them.

In contrast settings.yaml was a non-standard, KF specific solution. So I think getting rid of that in favor of standardizing on Kpt is preferable.

Bobgy replied:
Let's move the discussion to https://github.com/kubeflow/gcp-blueprints/issues/158.
This issue is not in scope for this PR.

Bobgy commented on management/README.md:
````diff
@@ -4,14 +4,90 @@ This directory contains the configuration needed to setup a management GKE clust
+1. Install the config connector:
+    ```bash
+    make apply-kcc
+    ```
````
I think we are trying to teach users what we are doing under the hood, so that they can continue to maintain this script on their own.

chensun replied:
That make sense.

Bobgy reviewed (commented):
Thanks @chensun for so much valuable feedback!

subodh101 commented on management/Makefile:
```diff
@@ -53,12 +62,14 @@ get-pkg:
 	kpt cfg set ./upstream/management gcloud.core.project $(PROJECT)
 
+.PHONY: hydrate
+hydrate: validate-values set-upstream-values
```
[CAN]
Also for the create context rule.

```
create-ctxt: validate-name-length validate-values
```

Or just better to merge the `validate-name-length` in the `validate-values`.

subodh101 replied:
This has been resolved.
https://github.com/kubeflow/gcp-blueprints/pull/153

subodh101 commented on management/Makefile:
```diff
@@ -85,25 +96,37 @@ hydrate-kcc:
 	kubectl wait -n configconnector-operator-system --for=condition=Ready pod --all --timeout 5m
-	# Wait for config connector pods are created
+	# Wait until config connector pods are created
+	# TODO(Bobgy): is there a better way than sleep?
```
[ASK] Can you create an issue? I will pick it up.
I think we will have to monitor the status of the operator. That will do the job here.

Bobgy replied:
Thanks!
Updated https://github.com/kubeflow/gcp-blueprints/issues/159

subodh101 commented on management/README.md:
```diff
@@ -4,14 +4,136 @@ This directory contains the configuration needed to setup a management GKE clust
 
 Please refer to the latest [docs](https://master.kubeflow.org/docs/gke/deploy/management-setup/)
 
+## Install Instructions
```
[ASK] Do we also need the tools version here?
Ref.: https://www.kubeflow.org/docs/gke/deploy/management-setup/#install-the-required-tools.

Bobgy replied:
I was only trying to add instructions that changed here (because it's easier to review instruction change + code together).
I'll merge the instructions to https://www.kubeflow.org/docs/gke/deploy/management-setup/#install-the-required-tools, once we are ready to release.

subodh101 commented on management/README.md:
```diff
@@ -4,14 +4,136 @@ This directory contains the configuration needed to setup a management GKE clust
+    --role=roles/owner
+gcloud --project=${PROJECT} iam service-accounts delete \
+    ${NAME}-cnrm-system@${PROJECT}.iam.gserviceaccount.com
+gcloud --project=${PROJECT} container clusters delete \
```
[ASK]
Once a management cluster is used to deploy kubeflow.
If we remove the associated resources and delete the management cluster. Will a user be able to upgrade or do operation on the same deployed kubeflow in future? by creating a new management cluster with the same configurations?

In my case, we keep the management cluster at size zero as it is not in use.

Bobgy replied:
Yes, that's possible.
https://cloud.google.com/config-connector/docs/how-to/managing-deleting-resources#acquiring_a_dataset

subodh101 replied:
Great! thank you!

subodh101 commented:
For the failing tests,

We need to add a zone variable "gcloud.compute.zone".
https://github.com/kubeflow/gcp-blueprints/blob/master/tests/validate_test.go#L58

We can get rid of `upstreamVars` as well.
Ref: https://github.com/kubeflow/testing/pull/759/files

Bobgy commented:
/retest

subodh101 commented:
Seems like new changes were not reflected from retest. Also, the test succeeded.
One check before JUnit results really becomes an issue.
/test all
/lgtm
Thank You!

subodh101 commented:
/hold
Please remove hold at your convenience.

Bobgy commented:
TestHydrate: validate_test.go:51: Run /google-cloud-sdk/bin/kpt cfg set instance mgmt-ctxt mgmt-ctxt:
        set 0 fields
    TestHydrate: validate_test.go:51: Run /google-cloud-sdk/bin/kpt cfg set instance name kf-test:
        set 9 fields
    TestHydrate: validate_test.go:51: Run /google-cloud-sdk/bin/kpt cfg set instance gcloud.core.project kubeflow-ci-deployment:
        set 6 fields
    TestHydrate: validate_test.go:51: Run /google-cloud-sdk/bin/kpt cfg set instance location us-east1:
        set 0 fields
    TestHydrate: validate_test.go:51: Run /google-cloud-sdk/bin/kpt cfg set instance email user@gmail.com:
        set 2 fields
    TestHydrate: validate_test.go:51: Run /google-cloud-sdk/bin/kpt cfg set upstream/manifests/gcp name kf-test:
        set 145 fields
    TestHydrate: validate_test.go:51: Run /google-cloud-sdk/bin/kpt cfg set upstream/manifests/gcp gcloud.core.project kubeflow-ci-deployment:
        set 84 fields
        set 3 fields
    TestHydrate: validate_test.go:51: Run /google-cloud-sdk/bin/kpt cfg set upstream/manifests/gcp gcloud.compute.zone us-east1-d:
        set 0 fields
    TestHydrate: validate_test.go:51: Run /google-cloud-sdk/bin/kpt cfg set upstream/manifests/gcp location us-east1:
        set 10 fields
    TestHydrate: validate_test.go:79: Run /usr/bin/make hydrate:
        Makefile:41: *** Either of MGMTCTXT, NAME, ZONE, LOCATION, PROJECT values not set.  Stop.

Bobgy commented:
It seems I need to manually merge master to get latest updates to the presubmit test

Bobgy commented:
I've addressed comments in this PR.
Therefore, I think we can get this in first.

Will send another PR to address https://github.com/kubeflow/gcp-blueprints/issues/158

Bobgy commented:
/unhold
